src/quake/dataset/generate_utils.py

```diff
@@ -0,0 +1,216 @@
+
+    Returns
+    -------
+        - x hit position of sh-ape=(tracks, [hits])
+        - y hit position of shape=(tracks, [hits])
+        - z hit position of shape=(tracks, [hits])
+        - hit energy of shape=(tracks, [hits])
+    """
+    rng = np.random.default_rng(seed=seed)
+    with uproot.open(name) as sig_root:
+        qtree = sig_root["qtree"]
+
+        xs = qtree["TrackPostX"].array()
+        ys = qtree["TrackPostY"].array()
+        zs = qtree["TrackPostZ"].array()
+        Es = qtree["TrackEnergy"].array()
+        tid = qtree["TrackID"].array()
+
+    if is_signal:
+        # concatenate the two b tracks (from two consecutive rows)
+        cat_fn = lambda arr: ak.concatenate([arr[::2], arr[1::2]], axis=1)
+        xs = cat_fn(xs)
+        ys = cat_fn(ys)
+        zs = cat_fn(zs)
+        Es = cat_fn(Es)
+        tid = cat_fn(tid)
+
+    Xs = ak.sum(xs[tid == 1] * Es[tid == 1], axis=1) / ak.sum(Es[tid == 1], axis=1)
+    Ys = ak.sum(ys[tid == 1] * Es[tid == 1], axis=1) / ak.sum(Es[tid == 1], axis=1)
+    Zs = ak.sum(zs[tid == 1] * Es[tid == 1], axis=1) / ak.sum(Es[tid == 1], axis=1)
+
+    normalize = (
+        lambda arr, shift, bw: arr - shift + rng.uniform(low=-bw, high=bw, size=1000)
+    )
+
+    xs = normalize(xs, Xs, geo.xbin_w)
+    ys = normalize(ys, Ys, geo.ybin_w)
+    zs = normalize(zs, Zs, geo.zbin_w)
+    return xs, ys, zs, Es
+
+
+def tracks2histograms(
+    xs: ak.Array, ys: ak.Array, zs: ak.Array, Es: ak.Array, geo: Geometry
+) -> ak.Array:
+    """
+    Compute energy histogram from track hit positions. This function converts
+    the simulated hit energy depositions to pixel images. The Geometry object,
+    passed as a parameter, controls the binning resolution and other
+    histogramming settings.
+
+    Parameters
+    ----------
+        - xs: x hit position of shape=(tracks, [hits])
+        - ys: y hit position of shape=(tracks, [hits])
+        - zs: z hit position of shape=(tracks, [hits])
+        - Es: hit energy of shape=(tracks, [hits])
+        - geo: detector geometry
+
+    Returns
+    -------
+        - sparse energy histogram of shape=()
+    """
+    logger.debug("Converting to histogram ...")
+    hists = []
+
+    # filter out of bin range hits
+    mx = np.logical_and(xs >= geo.xbins[0], xs < geo.xbins[-1])
+    my = np.logical_and(ys >= geo.ybins[0], ys < geo.ybins[-1])
+    mz = np.logical_and(zs >= geo.zbins[0], zs < geo.zbins[-1])
+    m = np.logical_and(np.logical_and(mx, my), mz)
+
+    for x, y, z, energy in zip(xs[m], ys[m], zs[m], Es[m]):
+        # digits start from 1
+        get_digit = lambda p1, p2: np.digitize(p1, p2) - 1
+        x_digits = get_digit(x.to_numpy(), geo.xbins)
+        y_digits = get_digit(y.to_numpy(), geo.ybins)
+        z_digits = get_digit(z.to_numpy(), geo.zbins)
+
+        yz_digits = y_digits * geo.nb_zbins + z_digits
+        shape = (geo.nb_xbins, geo.nb_ybins * geo.nb_zbins)
+        cutoff = 0.1
+        hist = sparse.csr_matrix(
```

Comment:
```suggestion
        hist = sparse.csr_matrix(
            (energy.to_numpy(), (x_digits, yz_digits)), shape=shape
        )

        # thresholding
        cutoff = 0.1
        rows, cols = hist.nonzero()
        values = np.array(hist[rows, cols])[0]

        mask = values < cutoff
        if np.count_nonzero(mask):
            values[mask] = 0
            hist = sparse.csr_matrix(
                (values, (rows, cols)), shape=shape
            )

            hists.append(hist.reshape(1, -1))
```
This might be clearer. You want to shut down pixels with energy below 0.1, right?